### DISCUSSION TITLE
Can't figure out how to enable the language server for python

### DISCUSSION THREAD

**Discussion by nitanmarcel** [kakoune-lsp/kakoune-lsp#776] [General]
I just set up kakoune and kakoune-lsp but I can't get to make the server to run. When I type for example `import os` then try to use `os.path` it does not show any autocomplete.

# I'm installing kak-lsp via kak-bundle

```
bundle kak-lsp https://github.com/kak-lsp/kak-lsp %{
  # Configure here...
  hook global KakEnd .* lsp-exit
}

bundle-install-hook kak-lsp %{
  # Any shell code that needs to be run goes here...
  cargo install --locked --force --path .
}

bundle-cleaner kak-lsp %{
  rm ~/.cargo/bin/kak-lsp
}

hook global WinSetOption filetype=(rust|python|go|javascript|typescript|c|cpp) %{
    lsp-enable-window
    lsp-auto-hover-enable

    set-option window lsp_hover_anchor false
    set-option window lsp_hover_max_lines 10

    set-face window DiagnosticError red,default+i
    set-face window DiagnosticHint yellow,default+i
    set-face window DiagnosticWarning default+u

    set-option global lsp_cmd "kak-lsp -vvv -c ~/.config/kak-lsp/kak-lsp.toml -s %val{session}"
    set-option global lsp_server_configuration pyls.configurationSources=["flake8"]

}
```

# and here is my kak-lsp.toml

```
snippet_support = true
verbosity = 2

[server]
timeout = 1800

[language_ids]
python = "python"

[language_server.python]
filetypes = ["python"]
roots = ["requirements.txt", "setup.py", ".git", ".hg", "Pipefile"]
command = "pyls"
offset_encoding = "utf-8"
```

# Logs

```
221│LSP:·ERRO·panic:·panicked·at·src/settings.rs:88:14:¬
222│called·`Option::unwrap()`·on·a·`None`·value¬
```

**Comment by nitanmarcel** [comment]
I found out thanks to another github repository. My final config is:

```
[language_server.pylsp]
filetypes = ["python"]
roots = ["requirements.txt", "setup.py", ".git", ".hg"]
command = "pylsp"
settings_section = "_"

[language_server.pylsp.settings._]
pylsp.plugins.jedi_completion.include_params = true
```

**Reply by krobelus** [reply]
> 221│LSP:·ERRO·panic:·panicked·at·src/settings.rs:88:14:¬
> 222│called·`Option::unwrap()`·on·a·`None`·value¬

Thanks for reporting, this was a recent regression.
I just fixed it in 589bd1f5 (Don't use server settings from new config if running off legacy kak-lsp.toml, 2024-09-17)

With that fix, your original `kak-lsp.toml` seems to work as expected.

**Comment by krobelus** [comment]
> bundle kak-lsp https://github.com/kak-lsp/kak-lsp %{
>   nop Configure here...
>   hook global KakEnd .* lsp-exit

this hook should be unnecessary as of latest master

> }
> 
> bundle-install-hook kak-lsp %{
>   nop Any shell code that needs to be run goes here...
>   cargo install --locked --force --path .
> }
> 
> bundle-cleaner kak-lsp %{
>   rm ~/.cargo/bin/kak-lsp
> }
> 
> hook global WinSetOption filetype=(rust|python|go|javascript|typescript|c|cpp) %{
>     lsp-enable-window
>     lsp-auto-hover-enable

'lsp-auto-hover-enable' will add a globally-scoped hook.
This means that whenever you open a file, you add another one of these hooks, leading to lots of redundant requests,
A cheesy workaround would be to call `lsp-auto-hover-disable` before this line.
I guess we should give it a scope argument to prevent such misconfiguration.

>     set-option global lsp_cmd "kak-lsp -vvv -c ~/.config/kak-lsp/kak-lsp.toml -s %val{session}"

This line should be redundant as of latest master (well except for the `-vvv` but that is obsoleted by `lsp_debug` which applies immediately).
I guess few people will notice the config change unless we actually break things.
As of today's master, the `-s` argument prevents us from doing proper cleanup on `rename-session`. Still figuring out what to do here..

>     set-option global lsp_server_configuration pyls.configurationSources=["flake8"]

Note that this option and kak-lsp.toml will be deprecated in the next release
but we're still figuring out the exact UI of the new option-based config.
So it's fine to use it but just keep in mind that the docs in master no longer mentions `lsp_server_configuration` or `kak-lsp.toml`.
